### PR TITLE
Bug 1208003 - [TV 2.5] Implement the functionalities of the TV remote control app

### PR DESCRIPTION
**Pull request by luke-chang**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=4df1b9c51a2541fdb5349bd2910c391703a38c5a

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=fd6c656b59d0faf1adf8d781cb8938eaca4d15a3

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=bb161326bb34d4d688e0150796f8a26422f2fbf3

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=35c59afeed1c250368954f070f933e7a6068ecac

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=fc441c1abe7f0b62d9ff24bf0e71d1f485e24e1f

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=fd65d203cc33d9048b77789b3e47c2e93d83e240

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c52ae7bcccb4873cb38b5c89ee79c3028b81a7ff

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e67b62b084701d49fefc9d81308a2f49c3bca7e7

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e5978def14d0ed539aadede1afe4f5d47691b2e6

File: tv_apps/remote-control/js/app.js [hunk 1]
```diff
@@ -0,0 +1,144 @@
+/* global KeyNavigationAdapter, Section, Config */
+'use strict';
+
+(function(exports) {
+  var MAIN_SECTION = 'main-section';
+  var DEFAULT_ELEMENT = 'qrcode';
```
Comment (evelynhung) [inline tv_apps/remote-control/js/app.js]: The variable name didn't have too much meaning to me. Does it mean `DEFAULT_FOCUSED_ELEMENT`?

File: tv_apps/remote-control/js/app.js [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      setVisible('location', enabled && hasConnection);
+      setVisible('off-line-message', enabled && !hasConnection);
+      setVisible('disabled-message', !enabled);
+
+      var mainSection = this.sections[MAIN_SECTION];
+      if (enabled && hasConnection) {
+        mainSection.focus(document.getElementById(DEFAULT_ELEMENT));
+      } else {
+        mainSection.unfocus();
+        mainSection.focus();
```
Comment (evelynhung) [inline tv_apps/remote-control/js/app.js]: Why do we need to `unfocus()` and then `focus()` ? Please add comments here.

File: tv_apps/remote-control/js/config.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+'use strict';
+
+(function(exports) {
+  var PREFIX = 'remote-control.';
+  var SETTINGS = {
+    'enabled': false,
+    'pairing-required': false,
+    'authorized-devices': null,
+    'server-ip': ''
+  };
```
Comment (evelynhung) [inline tv_apps/remote-control/js/config.js]: nit: I personally feel concatenating setting keys in this way is not readable.

File: tv_apps/remote-control/js/config.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+  var DEBUG = false;
+
+  function Config() {
+    this._ready = false;
```
Comment (evelynhung) [inline tv_apps/remote-control/js/config.js]: rename `_ready` to be a meaningful one.

File: tv_apps/remote-control/js/config.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+    _settingHandler: function(name, value) {
+      if (DEBUG) {
+        console.log('[Config] settings changed: ' + name + '=' + value);
+      }
+
+      this.settings[name] = value;
+      this.fire('changed', name, value);
+
+      if (!this._ready) {
```
Comment (evelynhung) [inline tv_apps/remote-control/js/config.js]: nit: it's better to have a comment here to describe the purpose of this code segment.

File: tv_apps/remote-control/js/app.js [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+      this.config = null;
+
+      this.keyNav.uninit();
+      this.keyNav = null;
+
+      this.activeSection = '';
+      this.sections = [];
+    },
+
+    _statusCheck: function() {
```
Comment (evelynhung) [inline tv_apps/remote-control/js/app.js]: I didn't see too many "check" action but a few status reset, we should rename this function to be something meaningful and add comments to describe its purpose.

File: tv_apps/remote-control/js/config.js (hunk 3 above)
Comment (evelynhung) [inline tv_apps/remote-control/js/config.js]: By reading more codes, I feel this concatenation may increase the effort of debugging and tracing code. Strongly suggest we always use complete name of keys.

File: tv_apps/remote-control/js/app.js [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+          break;
+        case 'save-config':
+          this.config.save();
+          // wait for settings change
+          setTimeout(this._switchSection.bind(this), 200);
+          break;
+      }
+    },
+
+    _handleBack: function() {
```
Comment (evelynhung) [inline tv_apps/remote-control/js/app.js]: We are simplifying page transition flow and assuming at most two layers of pages. Please add comments here as a warning.

File: tv_apps/remote-control/js/config.js [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        case 'option-pair-on':
+          settings[PREFIX + 'enabled'] = true;
+          settings[PREFIX + 'pairing-required'] = true;
+          break;
+        case 'option-pair-off':
+          settings[PREFIX + 'enabled'] = true;
+          settings[PREFIX + 'pairing-required'] = false;
+          break;
+        case 'option-disabled':
+          settings[PREFIX + 'enabled'] = false;
```
Comment (evelynhung) [inline tv_apps/remote-control/js/config.js]: I feel we should give `pairing-required` a status although it doesn't hurt. I vote for `true`.

File: tv_apps/remote-control/js/config.js [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+      if (canClear) {
+        button.removeAttribute('aria-hidden');
+      } else {
+        button.setAttribute('aria-hidden', true);
+      }
+
+      callback(checked);
+    },
+
+    check: function(id) {
```
Comment (evelynhung) [inline tv_apps/remote-control/js/config.js]: nit: `checkRadio`

File: tv_apps/remote-control/js/config.js [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+      for(var name in SETTINGS) {
+        SettingsListener.unobserve(PREFIX + name, this._observer[name]);
+      }
+      this.settings = [];
+      this._observer = [];
+      this._ready = false;
+    },
+
+    prepareOptions: function(callback) {
+      var checked;
```
Comment (evelynhung) [inline tv_apps/remote-control/js/config.js]: nit: `checkedOption`

File: tv_apps/remote-control/js/app.js [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+            .then((value) => {
+              if (confirm(value)) {
+                this.config.clearAuthorizedDevices();
+              }
+            });
+          break;
+        case 'save-config':
+          this.config.save();
+          // wait for settings change
+          setTimeout(this._switchSection.bind(this), 200);
```
Comment (evelynhung) [inline tv_apps/remote-control/js/app.js]: Let's remove this magic number and use callback in `save` function.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ad626b2f01679cdf6d508855bc130f5c50424702

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=98203239630b76dff78ee71918e69b248ccd223a

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7d21ed73e7bc7c73e62612354349cc6f22319ed8

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7a0acc4e2a0051bee3b53be800e9dd3e384dd5b6

File: tv_apps/remote-control/js/settings.js [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+      });
+    },
+
+    _settingHandler: function(name, value) {
+      if (this[name] !== value) {
+        this[name] = value;
+        this.fire('changed', name, value);
+      }
+
+      // Check whether all settings are retrieved at least once.
```
Comment (evelynhung) [inline tv_apps/remote-control/js/settings.js]: nit: remove `at least once`.

File: tv_apps/remote-control/js/app.js [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+  App.prototype = Object.create(Section.prototype);
+
+  // Section's Interface
+  App.prototype._readyToShow = function(previousId) {
+    this.updateStatus();
+
+    if (!previousId) {
+      this.focus(document.getElementById('qrcode'));
+    }
```
Comment (evelynhung) [inline tv_apps/remote-control/js/app.js]: how about `else` case?

File: tv_apps/remote-control/js/config_section.js [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    var elem = this.getFocusedElement();
+    switch(elem.id) {
+      case 'option-pair-off':
+      case 'option-pair-on':
+      case 'option-disabled':
+        this.checkRadio(elem.id);
+        break;
+      case 'clear-paired-devices':
```
Comment (evelynhung) [inline tv_apps/remote-control/js/config_section.js]: I'm confused by your interchanging term "paired device" and "authorized device". Could we unify them?

Comment (evelynhung) [inline tv_apps/remote-control/js/config_section.js]: I prefer "authorized".

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6394f0c958af03d10d9879766d2aba6f940bda6e

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=936d5b31e64fab647afc4d0b3373d3227d1a46ca

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=37b619b680eb88020fdddd3fac5f6393b98fa662

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=f99179bf9443becb77e64453431e9a55f0cc6af8

**Pull request merged**